src/expression/ekat_expression_base.hpp

```diff
@@ -0,0 +1,40 @@
+#ifndef EKAT_EXPRESSION_HPP
+#define EKAT_EXPRESSION_HPP
+
+#include <Kokkos_Core.hpp>
+
+namespace ekat {
+
+template<typename Derived>
+class Expression {
+public:
+
+  int num_indices () const { return cast().num_indices(); }
```

Comment:
I think this could be turned into a static fcn, as the number of indices used by an expression is fully determined by its type.

So far, this fcn is only used in the `evaluate` call, to make sure that the rank of the view passed to the evaluation fcn is compatible with the expression to evaluate. I may revisit this pattern in a follow-up pr...